_plugins/link-checker.rb

```diff
@@ -1,219 +1,376 @@
+        process(page)
+      end
+
+      # Process a Document as soon as its content is ready
+      Jekyll::Hooks.register :documents, :post_convert, priority: priority do |document|
+        process(document)
+      end
+
+      # Verify gathered links after Jekyll is done writing all its stuff
+      Jekyll::Hooks.register :site, :post_write, priority: priority do |site|
+        verify(site)
+      end
+
+      if @check_links
+        Jekyll.logger.info "LinkChecker: [Notice] Initialized successfully and will check #{msg[check_flag]}"
+      end
+      Jekyll.logger.info 'LinkChecker: [Notice] The build will fail if a dead link is found' if @should_build_fatally
+    rescue StandardError => e
+      Jekyll.logger.error 'LinkChecker: [Error] Failed to initialize Link Checker'
+      raise
+    end
   end
 
+  ##
   # Processes a Document or Page and adds the links to a collection
-  # It also checks for anchors that link to parts of the same page/doc
-  # return [void]
+  # It also checks for anchors to parts of the same page/doc
+
   def self.process(page)
+    return unless @check_links
     return if @excluded_paths.match(page.path)
 
     hrefs = page.content.scan(@href_matcher)
     hrefs.each do |(_, href)|
       relative_path = page.path[0] == '/' ? Pathname.new(page.path).relative_path_from(Dir.getwd) : page.path
 
-      if href.start_with? '#'
-        @failures << "##{href[1..]}, linked in ./#{relative_path}" if (page.content =~ /<[a-z0-9-]+[^>]+id="#{href[1..]}"/i).nil?
-      else
-        match = @base_url_matcher.match(href)
-        unless match.nil?
-          href = match[1]
+      if href.eql? '#'
+        next
+      elsif href.start_with? '#'
+        Jekyll.logger.info relative_path if (page.content =~ /<[a-z0-9-]+[^>]+(?:id|name)="#{href[1..]}"/i).nil?
+        if (page.content =~ /<[a-z0-9-]+[^>]+(?:id|name)="#{href[1..]}"/i).nil?
+          @failures << "Process:: ##{href[1..]}, linked in ./#{relative_path}"
         end
+      else
         @urls[href] = Set[] unless @urls.key?(href)
         @urls[href] << relative_path
       end
     end
   end
 
-  # Verifies the validity of all the destinations gathered in @urls
-  # return [void]
-  def self.verify(site)
-    if ENV.key?('JEKYLL_CHECK_EXTERNAL_LINKS')
-      @check_external_links = true
-      puts "LinkChecker: [Notice] Will verify external links"
-    end
+  ##
+  # Saves the collection as a JSON file
 
-    if ENV.key?('JEKYLL_FATAL_LINK_CHECKER')
-      @should_build_fatally = true
-      if ENV['JEKYLL_FATAL_LINK_CHECKER'] == '2'
-        @check_external_links = true
-        puts "LinkChecker: [Notice] The build will fail if any dead links are found"
-      else
-        puts "LinkChecker: [Notice] The build will fail if a dead internal link is found"
+  def self.verify(_site)
+    return unless @check_links
+
+    @base_url_matcher = %r{^#{@site.config["url"]}#{@site.baseurl}(/.*)$}.freeze
```

Comment:
Could we add a check for the case where a writer forgets a slash after the base URL? For example, `{{site.url}}{{site.baseurl}}opensearch/supported-field-types/range/`.

Reply:
It's a bit tricky, let's defer this to a future PR.